Sources/NIO/RecvByteBufferAllocator.swift

```diff
@@ -89,78 +66,50 @@ public struct AdaptiveRecvByteBufferAllocator: RecvByteBufferAllocator {
-            self.minIndex = minIndex + 1
-        } else {
-            self.minIndex = minIndex
-        }
-
-        let maxIndex = AdaptiveRecvByteBufferAllocator.sizeTableIndex(maximum)
-        if AdaptiveRecvByteBufferAllocator.sizeTable[maxIndex] > maximum {
-            self.maxIndex = maxIndex - 1
-        } else {
-            self.maxIndex = maxIndex
-        }
-
-        self.initial = initial
-        self.minimum = minimum
-        self.maximum = maximum
+        // We need to round all of these numbers to a power of 2. Initial will be rounded down,
+        // minimum down, and maximum up.
+        self.minimum = minimum.previousPowerOf2()
+        self.initial = min(initial, AdaptiveRecvByteBufferAllocator.maximumAllocationSize).previousPowerOf2()
+        self.maximum = min(maximum, AdaptiveRecvByteBufferAllocator.maximumAllocationSize).nextPowerOf2()
 
-        self.index = AdaptiveRecvByteBufferAllocator.sizeTableIndex(initial)
-        self.nextReceiveBufferSize = AdaptiveRecvByteBufferAllocator.sizeTable[index]
+        self.nextReceiveBufferSize = self.initial
         self.decreaseNow = false
     }
 
-    private static func sizeTableIndex(_ size: Int) -> Int {
-        var low: Int = 0
-        var high: Int = sizeTable.count - 1
+    public func buffer(allocator: ByteBufferAllocator) -> ByteBuffer {
+        return allocator.buffer(capacity: self.nextReceiveBufferSize)
+    }
 
-        repeat {
-            if high < low {
-                return low
-            }
-            if high == low {
-                return high
-            }
+    public mutating func record(actualReadBytes: Int) -> Bool {
+        precondition(self.nextReceiveBufferSize % 2 == 0)
+        precondition(self.nextReceiveBufferSize >= self.minimum)
+        precondition(self.nextReceiveBufferSize <= self.maximum)
 
-            // It's important to put (...) around as >> is executed before + in Swift while in Java it's the other way around (doh!)
-            let mid = (low + high) >> 1
+        var mayGrow = false
 
-            let a = sizeTable[mid]
-            let b = sizeTable[mid + 1]
+        // This right shift is safe: nextReceiveBufferSize can never be negative, so this will stop at 0.
+        let lowerBound = self.nextReceiveBufferSize &>> 1
 
-            if size > b {
-                low = mid + 1
-            } else if size < a {
-                high = mid - 1
-            } else if size == a {
-                return mid
-            } else {
-                return mid + 1
-            }
-        } while true
-    }
+        // Here we need to be careful with 32-bit systems: if maximum is Int32.max then any shift or multiply will overflow, which
+        // we don't want. Instead we check, and clamp to this current value if we overflow.
+        let (upperBoundCandidate, upperBoundOverflow) = self.nextReceiveBufferSize.multipliedReportingOverflow(by: 2)
```

Comment:
Why don't we make all these calculations on a type that's always big enough like `UInt32`? Then we can just shift and if the result is 0, then we overflowed?

Ie. `(UInt32(1) << 31) << 1 == 0`

Alternatively, we can just use `Int64` and also always shift and then check if `(self.nextReceiveBufferSize << 1) & Int64(UInt32.max) == 0` which then also would mean "too big".

Reply:
We absolutely can do. The reason I shied away from it is there is exactly one operation that's at any risk of overflowing, this one, and we can very easily clamp it instead. That seemed to me to be much simpler than hoisting into Int64 space and then doing math and clamping back down on the way out, with the fringe benefit that the compiler flags register already has the information we want, i.e. this operation degrades to `add, cmov`.

Reply:
Hmm, what I don't like about it is that it essentially runs different code on 32 and 64 bit platforms. The 64bit ones will never overflow.

Regarding the instruction selection:
- this is not a hot code path, we're going to allocate anyway
- I don't think (at least on 64bit) we'd generate worse code

But I don't mind really.

Reply:
> this is not a hot code path, we're going to allocate anyway

This is not quite true. This function is called once per `recv`, and this calculation is done to decide if we'll allocate. In the steady state, we never allocate, but we still always do this math.

I agree though that it runs different code on 32-and-64-bit platforms. That's somewhat unavoidable: at the edges of this type we need to work in `Int`-space, and so _somewhere_ we need to transform the input values from `Int` to whatever, and then transform back on the way out. Those operations are also different code on different platforms. That is an unavoidable fact of this API.

You're absolutely right that the code is no worse on Int64, that wasn't quite what I was saying: what I was saying is that this solution is no less optimal than doing the math in Int64. (Side note: onlimited 32-bit systems, none of which Swift supports, the original implementation may generate better code than doing the math in Int64-space if those systems have a limited number of 64-bit registers or no 64-bit instructions. This is not a factor in my decision making because to a first approximation we support none of those systems, but I want to bring it up for completeness.)

FWIW, here's the rewritten version, I'll merge which ever one you want:

```swift
        let upperBoundCandidate = Int(truncatingIfNeeded: Int64(self.nextReceiveBufferSize) &<< 1)
        let upperBound = upperBoundCandidate == 0 ? self.nextReceiveBufferSize : upperBoundCandidate
```

Reply:
> This is not quite true. This function is called once per `recv`, and this calculation is done to decide if we'll allocate. In the steady state, we never allocate, but we still always do this math.

fair enough, but `recv` means going to the kernel, which is even more expensive.


> I agree though that it runs different code on 32-and-64-bit platforms. That's somewhat unavoidable: at the edges of this type we need to work in `Int`-space, and so _somewhere_ we need to transform the input values from `Int` to whatever, and then transform back on the way out. Those operations are also different code on different platforms. That is an unavoidable fact of this API.

But we can do the math on a type where it _always_ works. And then transform the result into an `Int`. If we're over 2GB on 32bit, then yes this crashes but it also requires the _user_ to have configured an upper maximum of over 2GB.

If we want to be nice, we can just put an assertion where we take the users values and do `assert(userMax < Int.max)`, then they immediately get a crash when they configure the bogus value.


> You're absolutely right that the code is no worse on Int64, that wasn't quite what I was saying: what I was saying is that this solution is no less optimal than doing the math in Int64. (Side note: onlimited 32-bit systems, none of which Swift supports, the original implementation may generate better code than doing the math in Int64-space if those systems have a limited number of 64-bit registers or no 64-bit instructions. This is not a factor in my decision making because to a first approximation we support none of those systems, but I want to bring it up for completeness.)

Right, yeah, I completely ignore 32bit devices when it comes to perf. ByteBuffer for example is really bad on 32bit platforms because it won't fit in the existential...

> FWIW, here's the rewritten version, I'll merge which ever one you want:
> 
> ```swift
>         let upperBoundCandidate = Int(truncatingIfNeeded: Int64(self.nextReceiveBufferSize) &<< 1)
>         let upperBound = upperBoundCandidate == 0 ? self.nextReceiveBufferSize : upperBoundCandidate
> ```

I think that one's substantially nicer I think?

Reply:
> But we can do the math on a type where it always works. And then transform the result into an Int. If we're over 2GB on 32bit, then yes this crashes but it also requires the user to have configured an upper maximum of over 2GB.

Again, I agree, but I think that looks substantially _worse_ because we have 4 inputs to this function that all need to be duplicated into Int64 land before we compute on them, and two that need to be turned back into `Int` land on the way out. At a certain point we're drowning in integer conversions, each one of which has to be evaluated for correctness in 32-bit mode. That's why my proposed change was to modify exactly one line to do the dance: it reduces the cognitive pressure on the rest of the function. And, of course, any slip of the finger to use the original variable names screws you over again. :wink:

> I think that one's substantially nicer I think?

I think it is strictly worse, but you're reviewing.

Reply:
I already pushed the change.

Reply:
@Lukasa I'm okay with either version, my approval won't change whichever one you end up picking.